optuna/integration/__init__.py

```diff
@@ -1,6 +1,79 @@
-from optuna.integration.chainer import ChainerPruningExtension  # NOQA
-from optuna.integration.chainermn import ChainerMNStudy  # NOQA
-from optuna.integration.keras import KerasPruningCallback  # NOQA
-from optuna.integration.lightgbm import LightGBMPruningCallback  # NOQA
-from optuna.integration.tensorflow import TensorFlowPruningHook  # NOQA
-from optuna.integration.xgboost import XGBoostPruningCallback  # NOQA
+import os
+import sys
+from types import ModuleType
+from typing import Any  # NOQA
+
+from optuna.types import TYPE_CHECKING
+
+
+__all__ = [
+    'chainer',
+    'chainermn'
+    'keras',
+    'lightgbm',
+    'tensorflow',
+    'xgboost',
+    'ChainerMNStudy'
+    'ChainerPruningExtension',
+    'KerasPruningCallback',
+    'LightGBMPruningCallback',
+    'TensorFlowPruningHook',
+    'XGBoostPruningCallback',
+]
+
+
+if sys.version_info[0] == 2 or TYPE_CHECKING:
+    from optuna.integration.chainer import ChainerPruningExtension  # NOQA
+    from optuna.integration.chainermn import ChainerMNStudy  # NOQA
+    from optuna.integration.keras import KerasPruningCallback  # NOQA
+    from optuna.integration.lightgbm import LightGBMPruningCallback  # NOQA
+    from optuna.integration.tensorflow import TensorFlowPruningHook  # NOQA
+    from optuna.integration.xgboost import XGBoostPruningCallback  # NOQA
+else:
+    class _IntegrationModule(ModuleType):
+        """Module class that implements `optuna.integration` package.
+
+        This class applies lazy import under `optuna.integration`, where submodules are imported
+        when they are actually accessed. Otherwise, `import optuna` becomes much slower because it
+        imports all submodules and their dependencies (e.g., chainer, keras, lightgbm) all at once.
+        """
+
+        __path__ = [os.path.dirname(__file__)]
+
+        _modules = {
```

Comment:
[nits] I think this expression can be replaced with `_modules = set(_class_to_module.values())`. By this replacement, we can reduce a bit of redundant code.

Reply:
`__all__` may be able to auto-generate as well.

Reply:
Thanks for your pointing it out. Consolidated the import structure to a dictionary. 
https://github.com/pfnet/optuna/pull/334/files#diff-4f60853ac0648ba2dfecfa69ee7d64b5R9